DOCS-10559-aggregators-group-example-kt (#1519)
mulesoft/docs-connectors: aggregators/1.0/modules/ROOT/pages/aggregators-group-example.adoc

= Group-Based Aggregator Example - Mule 4

> The Aggregators module *Group based aggregator* scope enables you to aggregate elements in different groups according to a group ID. The following example shows how to configure this scope to aggregate up to `3` HTTP request calls per group based on the group ID sent in the request. Then, the application transforms the calls via DataWeave into random ID numbers. For each group that completes the aggregation of `3` elements, the application logs all the aggregated random ID number elements of the group.

Reviewer: ```suggestion
The Aggregators module *Group based aggregator* scope enables you to aggregate elements into groups by group ID. The following example shows how to configure this scope to aggregate up to `3` HTTP request calls per group based on the group ID sent in the request. The application than transforms the calls via DataWeave into random ID numbers. For each group that completes the aggregation of `3` elements, the application logs all the aggregated random ID number elements of the group.
```
It's probably just me, but take another look to be sure that there is no confusion between the group ID sent in the request and the random ID "number elements" (?) of the group. I see that these are syntactically `GroupId` and `Source ID`; I wonder if we should adopt this naming into the narrative.

}
----
[start=6]
. Drag the *Group based aggregator* scope to the right of *Transform message*.
. Set *Name* to `groupBasedAggregator`.
. Set *Content* to `payload`, which is the expression that defines what to aggregate: in this case, the transformed HTTP requests.
. Set *Group id* to `payload.GroupId`, which is the expression to evaluate for every new message received that gets the group ID where the element is aggregated.
. Set *Group size* to `3`, which is the maximum size of elements to aggregate to the group with the group ID resolved.
. Drag a *Logger* component into the *Incremental aggregation* route.
. In the *Logger* configuration screen, set *Message* to `Doing incremental aggregation step.` +
. Drag a *Logger* component into the *Aggregation complete* route.
. Set *Message* to `Aggregation complete`.
. Drag another *Logger* component to the right of the first *Logger* component in *Aggregation complete*.
. Set *Message* to the following expression, which returns all the aggregated elements:
+
[source,xml,linenums]
----
output application/json
---
payload
----
. Save the project.
. In *Package Explorer*, click *Run > Run As > Mule Application*.

> . Test the app by sending the curl command `http://localhost:8086/test?groupId=1` twice. +

Reviewer 2: I'm not sure if that's implied, but the curl is a POST (as configured in the XML below), so it will be like this
`curl -X POST http://localhost:8086/test\?groupId=1`

INFO  2021-06-14 16:34:51,409 [[MuleRuntime].uber.12: [aggregator-group-demo].aggregator-group-demoFlow.CPU_INTENSIVE @2c3478b5] [processor: aggregator-group-demoFlow/processors/1/route/1/processors/1; event: 96566ab0-cd47-11eb-8b82-f01898ad2638] org.mule.runtime.core.internal.processor.LoggerMessageProcessor: [
  {
    "Source Name": "group-based",
    "GroupId": "1",
    "Source ID": "0.5127046416902801"
  },
  {
    "Source Name": "group-based",
    "GroupId": "1",
    "Source ID": "0.20239307122656847"
  },
  {
    "Source Name": "group-based",
    "GroupId": "1",
    "Source ID": "0.7505233647798346"
  }
]
```
This example logger message shows that for the first two HTTP requests with group ID `1` and the third request with group ID `2`, the *Logger* component in the *Incremental aggregator* route logs the message `Doing incremental aggregation step.` +

> After you sent the fourth HTTP request with group ID `1`, the aggregation group `1` is complete because of the group size of `3` calls per group. The first *Logger* component in the *Aggregation complete* route logs the message `Aggregation complete`, and the second *Logger* component returns the aggregated payload elements of the group, which are requests transformed into random ID numbers.

Reviewer: ```suggestion
After you send the fourth HTTP request, which is the third request with group ID `1`, the aggregation group `1` is complete because of the group size of `3` calls per group. The first *Logger* component in the *Aggregation complete* route logs the message `Aggregation complete`, and the second *Logger* component returns the aggregated payload elements of the group, which are requests transformed into random ID numbers.
```

> If you send another two commands `http://localhost:8086/test?groupId=2` to aggregate calls to group `2`, the *Console* view returns the logger message:

Reviewer: Do we need to state here that doing so completes Group 2?
